### ISSUE TITLE
Code complete snippets for Open API Annotations

### ISSUE THREAD

**Issue by ericglau** [redhat-developer/quarkus-ls#246]
Support using code complete to insert code snippets/templates for Open API Annotations.

- Operation annotation with fields that can be filled in
![](https://github.com/ericglau/microprofile-language-server-demo/raw/master/images/OpenAPI-Snippet-Operation.gif)

- Parameter annotation with fields that can be filled in 
![](https://github.com/ericglau/microprofile-language-server-demo/raw/master/images/OpenAPI-Snippet-Parameter.gif)

See https://github.com/MicroShed/microprofile-language-server/blob/master/src/main/java/com/microprofile/lsp/internal/completion/MicroProfileCompletion.java

**Comment by angelozerr** [comment]
By using JDT LS, we benefit with completion on any annotation, so I wonder if we need this issue? The benefit with your completion snippet  feature is that it generates Parameter and description both (with standard JDT completion, it generates only Parameter). The cons with snippet that it doesn't generate the proper inport although standard JDT completion manage that.

But if you think it's important to provide this kind of snippet, I know JDT LS provides Eclipse template support. I never played with it and I don't know if we can trigger the template snippet before a class (and not everywhere).

@fbricon what do you think about that?

**Comment by ericglau** [comment]
MicroProfile OpenAPI has [many possible annotations](https://github.com/eclipse/microprofile-open-api/blob/master/spec/src/main/asciidoc/microprofile-openapi-spec.adoc#quick-overview-of-annotations), some which make use of multiple fields.

`@Parameter`with description is a simpler example.  For a more realistic example, see the `@APIResponses` annotation being used in [this code](https://github.com/OpenLiberty/guide-microprofile-openapi/blob/master/finish/src/main/java/io/openliberty/guides/inventory/InventoryResource.java).

With JDT LS, you would need to do code complete for the annotation and each of its fields separately.  For MicroProfile, it would be much easier for the developer if they can just insert a template with a default example as a starting point.

**Comment by angelozerr** [comment]
Thanks @ericglau for your explanation. I understand more now why you require this issue.

Please wait for I implement https://github.com/redhat-developer/vscode-quarkus/issues/119 The idea is to write a JSON snippet by using vscode format syntax. If you wish you could start writing that.

**Comment by ericglau** [comment]
Thanks @angelozerr.  We'll write up some JSON snippets.

**Comment by rzgry** [comment]
Started writing up some snippets here https://github.com/rzgry/microprofile-snippets/blob/master/snippets/snippets.json. @angelozerr Do these look like the correct format that the server will be expecting once redhat-developer/vscode-quarkus#119 is implemented?

Especially wondering if placeholders like `$1` etc. will work properly if the snippets are managed by the server

**Comment by angelozerr** [comment]
@rzgry your work is perfect and placeholders will work since it is managed by client side and it follows the LSP specification.

The only part which misses in your snippet is the:

 * `scope` to show the snippet according the completion offset (before the class name, etc). vscode uses textmate scope, so I suggest you that you do that. Once I will finish my work, teh scope will be computed by JDT LS with a custom JDT LS IDelegateCommand  (getASTScope for instance) to return the same value than textmate.

 * `context`: this is not in the part of vscode, but I would like to add it (for the moment do nothing about that). The idea is to show teh snippet only if openapi is on classpath for instance.

In other words:

 * please manage scope
 * please wait I finish my work to test it.

Thanks!

**Comment by rzgry** [comment]
@angelozerr do you know where I can find a list of the different possible scopes I can use?

**Comment by angelozerr** [comment]
Please forget my comment, I read too quickly https://code.visualstudio.com/docs/editor/userdefinedsnippets#_language-snippet-scope and I though it was textmate tokens which is not true. Sorry for the noise. 

But I would like to manage this usecase on server side. We will see that when we will integrate your snippet.

**Comment by angelozerr** [comment]
This issue is done without context filter. 

Snippets are written with vscode format (see https://github.com/redhat-developer/quarkus-ls/tree/master/microprofile.ls/com.redhat.microprofile.ls/src/main/resources/com/redhat/microprofile/snippets) and are executed on server side (Eclipse IDE can benefit with this feature for instance).

Please read documentation at https://github.com/redhat-developer/quarkus-ls/tree/master/microprofile.ls#code-snippets

But it misses context filter, I mean that @Operation snippet can be opened anywhere (in a method body for instance which is wrong).

Once I will implement https://github.com/redhat-developer/quarkus-ls/issues/265 @rzgry you will able to update your snippets with the proper context. It will give the capability to show open api snippets according to:

 * dependency, I mean only if project has microprofile-openapi dependency
 * Operation snippet will be available only for completion offset which is before class, method, etc

**Comment by angelozerr** [comment]
Fixed with https://github.com/redhat-developer/quarkus-ls/issues/265
